torch/nn/modules/loss.py

```diff
@@ -1121,6 +1125,91 @@ def forward(self, input: Tensor, target: Tensor) -> Tensor:
                                ignore_index=self.ignore_index, reduction=self.reduction)
 
 
+class CrossEntropyLossWithSoftLabels(_Loss):
+    r"""This criterion combines :class:`~torch.nn.LogSoftmax` and :class:`~torch.nn.KLDivLoss` in
+    one single class. It is useful when training a classification problem with `C` classes when
+    "soft labels" are required, such as when utilizing label smoothing or blended labels.
+
+    The `input` is expected to contain raw, unnormalized scores for each class and the `target`
+    should match the `input` shape and contain the target distribution over the classes.
+
+    The loss can be described as:
+
+    .. math::
+        L = l_{n,c} = y_{n,c} \cdot \left( \log y_{n,c} - \\
+        \log\left(\frac{\exp(x_{n,c})}{\sum_{i=1}^C \exp(x_{n,i})} \right) \right)
```

Comment:
I noticed that in #11959, @SsnL dropped the constant term. In [elegy, a jax nn library](https://github.com/poets-ai/elegy/blob/master/elegy/losses/categorical_crossentropy.py), the constant term is also dropped. I haven't been able to figure out how to read the TF source code for categorial_crossentropy yet, so idk what they're doing there.

Should we drop the constant term as well? This would change the implementation of this loss function from being a composition of LogSoftmax and KLDivLoss to being something else that is (1) more efficient to compute and (2) optimizes the same way (assuming your targets don't require grad...) but (3) isn't exactly the quantity users may expect.

Reply:
I think this is the TF source code for categorical_crossentropy and it looks like they drop the constant term as well: https://github.com/keras-team/keras/blob/0eb1954092e2110868e8ef06381cebac891e01e9/keras/backend.py#L4872

Reply:
Great question- definitely true that it'd be more efficient to drop the constant from an optimization perspective. Not sure what the impact is for 3 (user experience), since the one-hot case will still generally be equivalent between soft and hard cross losses (since the constant=0 for that case).

FWIW, SsnL doesn't think the constant should be dropped by default: https://github.com/pytorch/pytorch/issues/11959#issuecomment-861880668. I wonder if it makes sense for this PR to leave as-is and possibly add a flag for dropping the constant later on?

Reply:
Seems reasonable. Let's go ahead with this unless we get feedback from the other folks in the thread that they do want the constant dropped.